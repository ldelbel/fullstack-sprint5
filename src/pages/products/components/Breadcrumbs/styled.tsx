import styled from "styled-components";

export const StyledBreadcrumbs = styled.section`
  margin-bottom: 16px;

  nav {
    max-width: 1200px;
    margin: 0 auto;
  }

  .breadcrumbs__list {
    display: flex;
  }

  @media (max-width: 1200px) {
    .breadcrumbs__list {
      justify-content: center;
    }
  }
`;
