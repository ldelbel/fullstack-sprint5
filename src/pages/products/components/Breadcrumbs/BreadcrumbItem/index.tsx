import React from 'react';
import { StyledBreadcrumbItem } from "./styled";

interface IBreadcrumbItem {
  link: string;
  name: string;
}

export function BreadcrumbItem({ link, name }: IBreadcrumbItem) {
  return (
    <StyledBreadcrumbItem>
      {link ? (
        <>
          <a className="breadcrumbs__link" href={link}>
            {name}
          </a>
          <span className="breadcrumbs__item  breadcrumbs__separator">/</span>
        </>
      ) : (
        <span className="breadcrumbs__link">{name}</span>
      )}
    </StyledBreadcrumbItem>
  );
}