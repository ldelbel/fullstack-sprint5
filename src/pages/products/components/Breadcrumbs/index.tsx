import { useContext } from "react";
import CategoriesContext from "../../../../contexts/CategoriesContext";
import { ICategoriesContext } from "../../../../types/interfaces";
import { BreadcrumbItem } from "./BreadcrumbItem";
import { StyledBreadcrumbs } from "./styled";

function Breadcrumbs() {
  const { categories } = useContext<ICategoriesContext>(CategoriesContext);

  return (
    <StyledBreadcrumbs>
      <nav>
        <ol className="breadcrumbs__list">
          {categories.current &&
            categories.current.map((c) => (
              <BreadcrumbItem key={c.id} link={c.link} name={c.name} />
            ))}
        </ol>
      </nav>
    </StyledBreadcrumbs>
  );
}

export default Breadcrumbs;
