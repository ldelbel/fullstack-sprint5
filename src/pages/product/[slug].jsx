import { useEffect, useState } from "react";
import Breadcrumbs from "../products/components/Breadcrumbs";
import { Shopcard } from "./components/ShopCard";
import { SizeOption } from "./components/SizeOption";
import { StyledProduct } from './styled';

export function ProductInfo({ location }) {
  const { image, name, price } = location.state;
  const [sizes, setSizes] = useState([]);
  const [selectedSize, setSelectedSize] = useState();

  function defineSizes() {
    const array = name.split(' ').map(e => Number(e));
    const filtered = array.filter(e => !isNaN(e));
    const dif = filtered[1] - filtered[0];
    let i = 2;
    while(i < dif) {
      filtered.push(filtered[0] + i);
      i += 2;
    }
    setSizes(filtered.sort((a,b) => a - b))
  }

  useEffect(() => defineSizes(), []);

  return (
    <StyledProduct>
      <Breadcrumbs />
      <div className='content'>
        <div className="img">
          <img src={`/${image}`} alt="" className="content__image" />
        </div>
        
        <div className="content__description">

          <h2>
            {name}
          </h2>

          <div className="sizes">

            <p>Selecione um tamanho: <span>{selectedSize}</span></p>

            <div className="sizes__options">
              {sizes && sizes.map(size => (
                <SizeOption size={size} selected={selectedSize} setSize={setSelectedSize} />
              ))}
            </div>

          </div>
          <Shopcard price={price} />

        </div>

      </div>
    </StyledProduct>
  )
}