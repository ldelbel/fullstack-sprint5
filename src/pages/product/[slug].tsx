import React, { ReactElement, useEffect, useState } from "react";
import { useHistory, useLocation } from "react-router-dom";
import { TProduct } from "../../types";
import Breadcrumbs from "../products/components/Breadcrumbs";
import { ShopCard } from "./components/ShopCard";
import { SizeOption } from "./components/SizeOption";
import { StyledProduct } from "./styled";

export function ProductInfo(): ReactElement<React.FC> {
  const { state } = useLocation<TProduct>();
  const { push } = useHistory();
  const [sizes, setSizes] = useState<number[]>([]);
  const [selectedSize, setSelectedSize] = useState<number>();

  function defineSizes() {
    const array = state.name.split(" ").map((e) => Number(e));
    const filtered = array.filter((e) => !isNaN(e));
    const dif = filtered[1] - filtered[0];
    let i = 2;
    while (i < dif) {
      filtered.push(filtered[0] + i);
      i += 2;
    }
    setSizes(filtered.sort((a, b) => a - b));
  }

  useEffect(() => {
    if (!state) {
      push("/404");
      return;
    }
    defineSizes();

    // eslint-disable-next-line
  }, []);

  return state ? (
    <StyledProduct>
      <Breadcrumbs />
      <div className="content">
        <div className="img">
          <img src={`/${state.image}`} alt="" className="content__image" />
        </div>

        <div className="content__description">
          <h2>{state.name}</h2>

          <div className="sizes">
            <p>
              Selecione um tamanho: <span>{selectedSize}</span>
            </p>

            <div className="sizes__options">
              {sizes &&
                sizes.map((size) => (
                  <SizeOption
                    key={size}
                    size={size}
                    selected={selectedSize}
                    setSize={setSelectedSize}
                  />
                ))}
            </div>
          </div>
          <ShopCard price={state.price} />
        </div>
      </div>
    </StyledProduct>
  ) : (
    <div className="loading"></div>
  );
}
