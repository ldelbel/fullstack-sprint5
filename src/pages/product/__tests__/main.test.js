import React from "react";
import { render } from "@testing-library/react";
import { ProductInfo } from "../[slug]";
import { useLocation, MemoryRouter, Route } from "react-router-dom";
import { act } from "react-dom/test-utils";
import Breadcrumbs from "../../products/components/Breadcrumbs";

jest.mock("../../products/components/Breadcrumbs");
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: jest.fn(),
}));

beforeEach(() => {
  Breadcrumbs.mockImplementation(() => <div>Breadcrumbs</div>);
});

const state = {
  sku: 1,
  image: "assets/c1.webp",
  name: "Jaqueta Puffer Juvenil Com Capuz Super Mario Branco Tam 10 a 18",
  price: "199,90",
};

describe("ProductInfo component", () => {
  it("should render without crashing", () => {
    useLocation.mockImplementation(() => ({
      state,
    }));

    const { getByText } = render(
      <MemoryRouter initialEntries={["/jaqueta"]}>
        <ProductInfo />
      </MemoryRouter>
    );

    expect(getByText("Selecione um tamanho:")).toBeInTheDocument();
    useLocation.mockClear();
  });

  it("should get from /invalid to home on click", async () => {
    useLocation.mockImplementation(() => ({}));

    let testLocation;

    await act(async () => {
      render(
        <MemoryRouter initialEntries={["/jaqueta"]}>
          <ProductInfo />
          <Route
            path="*"
            render={({ location }) => {
              testLocation = location;
              return null;
            }}
          />
        </MemoryRouter>
      );
    });

    expect(testLocation.pathname).toBe("/404");
  });

  it("should match snapshot", () => {
    useLocation.mockImplementation(() => ({
      state,
    }));

    const { container } = render(
      <MemoryRouter initialEntries={["/jaqueta"]}>
        <ProductInfo />
      </MemoryRouter>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
