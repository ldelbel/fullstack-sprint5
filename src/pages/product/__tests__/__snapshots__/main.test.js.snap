// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductInfo component should match snapshot 1`] = `
<div
  class="sc-dlnjwi jrnRLM"
>
  <div>
    Breadcrumbs
  </div>
  <div
    class="content"
  >
    <div
      class="img"
    >
      <img
        alt=""
        class="content__image"
        src="/assets/c1.webp"
      />
    </div>
    <div
      class="content__description"
    >
      <h2>
        Jaqueta Puffer Juvenil Com Capuz Super Mario Branco Tam 10 a 18
      </h2>
      <div
        class="sizes"
      >
        <p>
          Selecione um tamanho: 
          <span />
        </p>
        <div
          class="sizes__options"
        >
          <button
            class="sc-gtsrHT lehJRh"
            data-testid="option"
          >
            10
          </button>
          <button
            class="sc-gtsrHT lehJRh"
            data-testid="option"
          >
            12
          </button>
          <button
            class="sc-gtsrHT lehJRh"
            data-testid="option"
          >
            14
          </button>
          <button
            class="sc-gtsrHT lehJRh"
            data-testid="option"
          >
            16
          </button>
          <button
            class="sc-gtsrHT lehJRh"
            data-testid="option"
          >
            18
          </button>
        </div>
      </div>
      <div
        class="sc-bdnxRM frziFI"
      >
        <p>
          Vendido e entregue por 
          <span>
            RCHLO
          </span>
        </p>
        <div
          class="prices"
        >
          <strong>
            R$ 
            199,90
          </strong>
          <div>
            <span>
              Cartão de crédito 1x de R$ 
              199,90
            </span>
            <span>
              Cartão de crédito 3x de R$ 
              66,63
            </span>
          </div>
        </div>
        <button
          class="addBtn"
        >
          ADICIONAR À SACOLA
        </button>
        <div
          class="deliver"
        >
          <div
            class="deliver__info"
          >
            <svg
              color="#333"
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              style="color: rgb(51, 51, 51);"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="13"
                width="15"
                x="1"
                y="3"
              />
              <polygon
                points="16 8 20 8 23 11 23 16 16 16 16 8"
              />
              <circle
                cx="5.5"
                cy="18.5"
                r="2.5"
              />
              <circle
                cx="18.5"
                cy="18.5"
                r="2.5"
              />
            </svg>
            <span>
              Consulte o frete
            </span>
            <input
              id="CEP"
              name="CEP"
              pattern="\\\\d{5}-\\\\d{3}"
              placeholder="______-___"
              required=""
            />
            <button>
              OK
            </button>
          </div>
          <p>
            Não sei meu cep
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
